python_binding/src/byte_stream.rs

```diff
@@ -0,0 +1,254 @@
+    stream: ScopedStream,
+    runtime_handle: Handle,
+    writer: ByteWriter,
+    reader: ByteReader,
+}
+
+// The amount of time the python api will wait for the underlying operation to be completed.
+const TIMEOUT_IN_SECONDS: u64 = 120;
+
+#[cfg(feature = "python_binding")]
+#[pymethods]
+impl ByteStream {
+    ///
+    /// Write a byte array into the Pravega Stream using a byte writer. The data is not persisted
+    /// on the server side when this method returns, user should call a flush to ensure all data has
+    /// been acknowledged by the server.
+    ///
+    /// ```
+    /// import pravega_client;
+    /// manager=pravega_client.StreamManager("tcp://127.0.0.1:9090")
+    /// // lets assume the Pravega scope and stream are already created.
+    /// byte_stream=manager.create_byte_stream("scope", "stream")
+    ///
+    /// byte_stream.write(b"bytes")
+    /// byte_stream.flush();
+    /// ```
+    ///
+    #[text_signature = "($self, byte_array)"]
+    #[args(byte_array)]
+    pub fn write(&mut self, byte_array: &[u8]) -> PyResult<usize> {
+        trace!("Writing a byte array to stream {:?}", self.stream);
+        match self.writer.write(byte_array) {
+            Ok(bytes_written) => Ok(bytes_written),
+            Err(e) => Err(exceptions::PyOSError::new_err(format!(
+                "Error while writing into ByteStream {:?}",
+                e
+            ))),
+        }
+    }
+
+    ///
+    /// This is is used to flush all the data to be persisted on the Pravega server side.
+    ///
+    /// ```
+    /// import pravega_client;
+    /// manager=pravega_client.StreamManager("tcp://127.0.0.1:9090")
+    /// // lets assume the Pravega scope and stream are already created.
+    /// byte_stream=manager.create_byte_stream("scope", "stream")
+    ///
+    /// byte_stream.write(b"bytes")
+    /// byte_stream.flush();
+    /// ```
+    ///
+    #[text_signature = "($self)"]
+    pub fn flush(&mut self) -> PyResult<()> {
+        info!("Flush all data into Pravega Stream {:?}", self.stream);
+        match self.writer.flush() {
+            Ok(()) => Ok(()),
+            Err(e) => Err(exceptions::PyOSError::new_err(format!(
+                "Error while flushing data into ByteStream {:?}",
+                e
+            ))),
+        }
+    }
+
+    ///
+    /// This is used to read bytes from a Pravega Stream into the specified buffer.
+    /// ```
+    /// import pravega_client;
+    /// manager=pravega_client.StreamManager("tcp://127.0.0.1:9090")
+    /// // lets assume the Pravega scope and stream are already created.
+    /// byte_stream=manager.create_byte_stream("scope", "stream")
+    ///
+    /// byte_stream.write(b"bytes")
+    /// byte_stream.flush();
+    /// buf = bytearray(5)
+    /// byte_stream.readinto(buf)
+    /// buf
+    /// bytearray(b'bytes')
+    /// ```
+    ///
+    #[text_signature = "($self, byteArray)"]
+    #[args(byteArray)]
+    pub fn readinto(&mut self, buf: &PyByteArray) -> PyResult<usize> {
```

Comment:
Why do we name it `readinto` instead of `read`, same naming convention as `tell`?

Reply:
Yeah, I followed the Python API pattern here. ref: https://docs.python.org/3/library/io.html#io.RawIOBase.readinto